R/range_search.R

```diff
@@ -0,0 +1,165 @@
+#' @title Range Search
+#'
+#' @description
+#' An implementation of range search with single-tree and dual-tree algorithms. 
+#' Given a set of reference points and a set of query points and a range, this
+#' can find the set of reference points within the desired range for each query
+#' point, and any trees built during the computation can be saved for reuse with
+#' future range searches.
+#'
+#' @param input_model File containing pre-trained range search model.
+#' @param leaf_size Leaf size for tree building (used for kd-trees, vp trees, random
+#'   projection trees, UB trees, R trees, R* trees, X trees, Hilbert R trees, R+
+#'   trees, R++ trees, and octrees).  Default value "20".
+#' @param max Upper bound in range (if not specified, +inf will be used.  Default
+#'   value "0".
+#' @param min Lower bound in range.  Default value "0".
+#' @param naive If true, O(n^2) naive mode is used for computation.  Default value
+#'   "FALSE".
+#' @param query File containing query points (optional).
+#' @param random_basis Before tree-building, project the data onto a random
+#'   orthogonal basis.  Default value "FALSE".
+#' @param reference Matrix containing the reference dataset.
+#' @param seed Random seed (if 0, std::time(NULL) is used).  Default value "0".
+#' @param single_mode If true, single-tree search is used (as opposed to dual-tree
+#'   search).  Default value "FALSE".
+#' @param tree_type Type of tree to use: 'kd', 'vp', 'rp', 'max-rp', 'ub', 'cover',
+#'   'r', 'r-star', 'x', 'ball', 'hilbert-r', 'r-plus', 'r-plus-plus', 'oct'. 
+#'   Default value "kd".
+#' @param verbose Display informational messages and the full list of parameters and
+#'   timers at the end of execution.  Default value "FALSE".
+#'
+#' @return A list with several components:
+#' \item{distances_file}{File to output distances into.  Default value "".}
+#' \item{neighbors_file}{File to output neighbors into.  Default value "".}
```

Comment:
This is another "weird" binding.  In C++, range search doesn't return an `arma::mat`; it can return a different number of results for each point in the input matrix, so actually it has to return a `std::vector<std::vector<size_t>>`.  For other languages, I've simply not added the range search binding for this reason.  Maybe that is the right thing to do for R too?